pkg/mcs/resource_manager/server/token_bukets.go

```diff
@@ -45,100 +52,229 @@ type GroupTokenBucket struct {
 	GroupTokenBucketState `json:"state,omitempty"`
 }
 
+func (gtb *GroupTokenBucket) setState(state *GroupTokenBucketState) {
+	gtb.mu.Lock()
+	defer gtb.mu.Unlock()
+
+	gtb.Tokens = state.Tokens
+	gtb.LastUpdate = state.LastUpdate
+	gtb.Initialized = state.Initialized
+	log.Info("update group token bucket state", zap.Any("state", state))
+}
+
+// TokenSlot is used to split a token bucket into multiple slots to
+// server different clients within the same resource group.
+type TokenSlot struct {
+	// settings is the token limit settings for the slot.
+	settings *rmpb.TokenLimitSettings
+	// assignTokens is the number of tokens in the slot.
+	assignTokens     float64
+	lastAssignTokens float64
+	requireTokensSum float64
+	lastUpdate       *time.Time
+}
+
 // GroupTokenBucketState is the running state of TokenBucket.
 type GroupTokenBucketState struct {
-	Tokens      float64    `json:"tokens,omitempty"`
+	mu     sync.Mutex
+	Tokens float64 `json:"tokens,omitempty"`
+	// ClientUniqueID -> TokenSlot
+	tokenSlots                 map[uint64]*TokenSlot
+	clientConsumptionTokensSum float64
+
 	LastUpdate  *time.Time `json:"last_update,omitempty"`
 	Initialized bool       `json:"initialized"`
 	// settingChanged is used to avoid that the number of tokens returned is jitter because of changing fill rate.
 	settingChanged bool
 }
 
 // Clone returns the copy of GroupTokenBucketState
-func (s *GroupTokenBucketState) Clone() *GroupTokenBucketState {
+func (gts *GroupTokenBucketState) Clone() *GroupTokenBucketState {
+	tokenSlots := make(map[uint64]*TokenSlot)
+	for id, tokens := range gts.tokenSlots {
+		tokenSlots[id] = tokens
+	}
+	var lastUpdate *time.Time
+	if gts.LastUpdate != nil {
+		newLastUpdate := *gts.LastUpdate
+		lastUpdate = &newLastUpdate
+	}
 	return &GroupTokenBucketState{
-		Tokens:      s.Tokens,
-		LastUpdate:  s.LastUpdate,
-		Initialized: s.Initialized,
+		Tokens:                     gts.Tokens,
+		LastUpdate:                 lastUpdate,
+		Initialized:                gts.Initialized,
+		tokenSlots:                 tokenSlots,
+		clientConsumptionTokensSum: gts.clientConsumptionTokensSum,
+	}
+}
+
+func (gts *GroupTokenBucketState) cleanupAssignTokenSum() {
+	gts.clientConsumptionTokensSum = 0
+	for _, slot := range gts.tokenSlots {
+		slot.requireTokensSum = 0
+	}
+}
+
+func (gts *GroupTokenBucketState) balanceSlotTokens(
+	clientUniqueID uint64,
+	settings *rmpb.TokenLimitSettings,
+	consumptionToken float64) {
+	now := time.Now()
+	if slot, ok := gts.tokenSlots[clientUniqueID]; !ok {
+		gts.cleanupAssignTokenSum()
+		gts.tokenSlots[clientUniqueID] = &TokenSlot{lastUpdate: &now}
+	} else {
+		slot.lastUpdate = &now
+		slot.requireTokensSum += consumptionToken
+		gts.clientConsumptionTokensSum += consumptionToken
+
+		if gts.clientConsumptionTokensSum >= maxAssignTokens {
+			gts.cleanupAssignTokenSum()
+		}
+	}
+
+	evenRatio := 1 / float64(len(gts.tokenSlots))
+	if settings.GetBurstLimit() < 0 {
```

Comment:
Why not do this check at the very first beginning?

Reply:
Because the first thing we need to check is: whether the client is new and whether exceeds the max limit for token size
and then check the burst limit to even fill rate.
If we put this check at the beginning when the client is new need to even again